### PR TITLE
improve tenant extraction from x509 cert

### PR DESCRIPTION
**Pull request by pdgetrf**
#### Issues fixed:
- Missing checks for edge cases
- A bug that could cause index out of bound with invalid input
- No unit testing

#### How fixed:
Added stricter checks for edge cases, fixed the bug and add full-fledged unit tests

#### Tests
Also tested e2e in local environment with tenant creation, role assignment, and rbac accesses.

#### Notes
- Make update was also run but didn't generate additional changes.
- TDD is fun :) 

Thanks @chenqianfzh for reporting this issue.

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go [hunk 1]
```diff
@@ -185,48 +185,51 @@ func DefaultVerifyOptions() x509.VerifyOptions {
 
 // CommonNameUserConversion builds user info from a certificate chain using the subject's CommonName
 var CommonNameUserConversion = UserConversionFunc(func(chain []*x509.Certificate) (*authenticator.Response, bool, error) {
```
Comment (chenqianfzh) [inline staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go:187]: "error", the last value of return, is always set to nil.  Shall we let convertToUserInfo() return some error message and the error is popped up in the call stack?

File: staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go [hunk 2]
```diff
@@ -185,48 +185,51 @@ func DefaultVerifyOptions() x509.VerifyOptions {
...
-		return &authenticator.Response{
-			User: &user.DefaultInfo{
-				Name:   chain[0].Subject.CommonName,
-				Groups: chain[0].Subject.Organization,
+	tenant := ""
+	if len(subject.Organization) > 0 {
+		parts := strings.Split(subject.Organization[0], ":")
+		if parts[0] == "tenant" {
+			// new format (tenant string from Organization)
+			if len(subject.Organization) > 1 {
```
Comment (chenqianfzh) [inline staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go:206]: You mean "len(parts) > 1", right?

File: staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go [hunk 3]
```diff
@@ -185,48 +185,51 @@ func DefaultVerifyOptions() x509.VerifyOptions {
...
+				// with new format, tenant comes from
+				// Organization and only one is allowed
+				return nil, false
+			}
+			numOfParts := len(parts)
+			if numOfParts == 2 {
```
Comment (chenqianfzh) [inline staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go:212]: The cases that the numOfParts is 1 is not handled. So if subject.Organization[0]=="tenant", we will get a userInfo where the tenant is ""

Comment (pdgetrf) [reply staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go:212]: that's the current defined behavior. see test case "new format, no colon tenant in organization"

File: staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go [hunk 4]
```diff
@@ -185,48 +185,51 @@ func DefaultVerifyOptions() x509.VerifyOptions {
...
+			}, true
+		}
 	}
 
-	// new format uses both /O and /OU
-	// one and only tenant per user
-	if len(chain[0].Subject.Organization) != 1 {
-		return nil, false, nil
+	// backward compatible, old format (tenant string from CN)
+	parts := strings.Split(subject.CommonName, ":")
```
Comment (chenqianfzh) [inline staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go:226]: Need to check the tenant is not empty. If CN is ":abc", the tenant might be empty.

Comment (pdgetrf) [reply staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go:226]: see test case "old format, nothing but a colon in CN"

File: staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go (hunk 1 above)
Comment (pdgetrf) [reply staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go:187]: good call. adding.

File: staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go [hunk 5]
```diff
@@ -185,48 +185,51 @@ func DefaultVerifyOptions() x509.VerifyOptions {
...
-			tenant = parts[0]
-		}
-		return &authenticator.Response{
-			User: &user.DefaultInfo{
-				Name:   chain[0].Subject.CommonName,
-				Groups: chain[0].Subject.Organization,
+	tenant := ""
+	if len(subject.Organization) > 0 {
+		parts := strings.Split(subject.Organization[0], ":")
+		if parts[0] == "tenant" {
```
Comment (chenqianfzh) [inline staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go:204]: if O="tenant" (no ":" in the end), it should be identified as an old format. But the current logic will treat it as a new format. 

So I think line 204 should be len(parts) >1 && parts[0] == "tenant"

File: staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go (hunk 3 above)
Comment (chenqianfzh) [inline staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go:212]: the case of numOfParts == 1 is not handled. However, if line 204 is changed as suggested, no change is needed here.

File: staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go [hunk 6]
```diff
@@ -185,48 +185,51 @@ func DefaultVerifyOptions() x509.VerifyOptions {
...
+				// Organization and only one is allowed
+				return nil, false
+			}
+			numOfParts := len(parts)
+			if numOfParts == 2 {
+				tenant = parts[1]
+			} else if numOfParts > 2 {
+				return nil, false
+			}
+			return &user.DefaultInfo{
```
Comment (chenqianfzh) [inline staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go:217]: Need to check the tenant value is not empty.

Comment (pdgetrf) [reply staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go:217]: tenant value being empty will fail any rbac check. this code here is only responsible to extract.

File: staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go (hunk 5 above)
Comment (pdgetrf) [reply staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go:204]: I think O="tenant" should be treated as the new format.

File: staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go (hunk 3 above)
Comment (pdgetrf) [reply staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go:212]: the case of numOfParts == 1 is handled implicitly as an "else" case of the "if" condition on line 212. this is on purpose.

File: staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509_test.go [hunk 7]
```diff
@@ -848,6 +849,238 @@ func TestX509Verifier(t *testing.T) {
...
+				Name: "abc:def:ghi",
+				Groups: []string{
+					"xyz",
+				},
+				Tenant: "abc",
+			},
+			shouldSucceed: true,
+		},
+		"old format, nothing but a colon in CN": {
+			input: pkix.Name{
```
Comment (chenqianfzh) [inline staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509_test.go:933]: Do we need to support such a CN, just a ":"?

File: staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509_test.go [hunk 8]
```diff
@@ -848,6 +849,238 @@ func TestX509Verifier(t *testing.T) {
...
+				OrganizationalUnit: []string{
+					"xyz",
+				},
+			},
+			expectUserInfo: &user.DefaultInfo{
+				Name: "abc",
+				Groups: []string{
+					"xyz",
+				},
+				Tenant: "",
```
Comment (chenqianfzh) [inline staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509_test.go:1003]: I don't think we should expect a valid cert will give a tenant of "".

File: staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509_test.go (hunk 7 above)
Comment (pdgetrf) [reply staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509_test.go:933]: I prefer to not add any rule as to what should or shouldn't be in a CN.

File: staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509_test.go (hunk 8 above)
Comment (pdgetrf) [reply staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509_test.go:1003]: this code is not to decide whether a cert is valid or not. it only tries to extract tenant string.

**Comment by pdgetrf** [comment]
> Guess we should have some offline discussion on what is a valid Arktos cert. The code allows a the tenant to be empty, which looks insecure to me.

it's not this code's job to perform authentication. an empty tenant will be denied by rbac when tenants do not match.

**Comment by pdgetrf** [comment]
> Guess we should have some offline discussion on what is a valid Arktos cert. The code allows a the tenant to be empty, which looks insecure to me.

another way to look at this is this code is in the authentication part. this part is to verify the user is what he/she claims he/she is. the tenant being empty or not should be enforced by the authorization step which determines what the user from his/her tenant can or cannot do. if we really want, a check can be added to authorization to deny all request from an empty tenant.

File: staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go (hunk 1 above)
Comment (pdgetrf) [reply staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go:190]: fixed. added error and changed tests accordingly

File: staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go (hunk 6 above)
Comment (chenqianfzh) [reply staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go:221]: Make sense. Agree.

File: staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go (hunk 3 above)
Comment (chenqianfzh) [reply staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go:216]: Got it. It is better to add a comment here to explain the case that O="tenant" is a valid case. It will help the maintenance of the code.

File: staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go (hunk 5 above)
Comment (chenqianfzh) [reply staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go:208]: I see your point.

File: staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go (hunk 3 above)
Comment (pdgetrf) [reply staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go:216]: will do

File: staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go [hunk 9]
```diff
@@ -185,48 +188,52 @@ func DefaultVerifyOptions() x509.VerifyOptions {
...
-	if len(chain[0].Subject.Organization) > 0 {
-		parts := strings.Split(chain[0].Subject.Organization[0], ":")
-		if len(parts) > 0 && parts[0] == "tenant" {
-			isNewformat = true
-			tenant = parts[1]
-		}
+func convertToUserInfo(subject pkix.Name) (*user.DefaultInfo, bool, error) {
+	errTenantExtraction := errors.New(ErrTenantExtraction)
+	if len(subject.CommonName) == 0 {
+		return nil, false, errTenantExtraction
```
Comment (chenqianfzh) [inline staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go:203]: Is errors.New("CommonName can not be empty") more descriptive?

**Comment by chenqianfzh** [comment]
> > Guess we should have some offline discussion on what is a valid Arktos cert. The code allows a the tenant to be empty, which looks insecure to me.
> 
> another way to look at this is this code is in the authentication part. this part is to verify the user is what he/she claims he/she is. the tenant being empty or not should be enforced by the authorization step which determines what the user from his/her tenant can or cannot do. we could enforce that user from an empty tenant cannot do anything and deny all such request but that falls into the authority of the authorization.

Good call. Thanks for the clarification.

File: staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go (hunk 9 above)
Comment (pdgetrf) [reply staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go:203]: I'll split the error message for each specific cases.

Comment (pdgetrf) [reply staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go:203]: fixed
